text/0000-use-mod-over-const.md

```diff
@@ -0,0 +1,65 @@
+- Feature Name: use-mod-over-const
+- Start Date: 2020-04-22
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+With stable support for attributes on modules the const workaround has a contender.
+
+# Motivation
+[motivation]: #motivation
+
+As noted in the rtfm::app documentation: 
+
+"This attribute must be applied to a const item of type ().
+The const item is effectively used as a *mod item*: its value must be a block that contains items commonly found in modules, like functions and static variables."
+
+Now with stable support in 1.42 for using actual mod the workaround using a const can be removed.
+
+# Detailed design
+[design]: #detailed-design
+
+[Stabilize attribute macros on inline modules #64273](https://github.com/rust-lang/rust/pull/64273) with milestone 1.42.
+
+Instead of `const APP: () = {` it is now possible to write just `mod APP {`.
+
+```
+//! examples/smallest.rs
+
+#![no_main]
+#![no_std]
+
+use panic_semihosting as _; // panic handler
+use rtfm::app;
+
+#[app(device = lm3s6965)]
+mod APP {}
+```
+
+However, since the macros operating on the mod remove the `mod` during code generation, the code within the module will see outside the module, which is not idiomatic.
```

Comment:
IMO `#[app]` should NOT break namespacing rules; that would be very surprising for end users.

Namely, this should not compile because `X` is not in scope:

``` rust
struct X;

#[app(device = "A")]
mod app {
    #[task]
    fn foo(_: foo::Context, input: X) { /* .. */ }
}
```

Instead you must write this, which is what you would write if there were no attributes

``` rust
struct X;

#[app(device = "A")]
mod app {
    use super::X; // <-

    #[task]
    fn foo(_: foo::Context, input: X) { /* .. */ }
}
```

Reply:
I have pushed a rough prototype implementation to the branches linked above.

By collecting the names of required resources/functions within various codegen modules, then by modifying rtfm-syntax to allow user entered `UseItem`s within the `mod app {}`.

I am not sure about modifying the App struct within rtfm-syntax, there may be neater ways to achieve the same result.

Feedback greatly appreciated!

All examples builds and running the test-suite locally passes.